connectors/sources/aws.py

```diff
@@ -3,116 +3,223 @@
 # or more contributor license agreements. Licensed under the Elastic License 2.0;
 # you may not use this file except in compliance with the Elastic License 2.0.
 #
-import os
-from hashlib import md5
 import asyncio
-from functools import partial
 import logging
+import os
+from functools import partial
+from hashlib import md5
 
 import aioboto3
+from aiobotocore.response import AioReadTimeoutError
+from aiohttp.client_exceptions import ServerTimeoutError
+from aiobotocore.config import AioConfig
 from aiobotocore.utils import logger as aws_logger
-
-from connectors.source import BaseDataSource
+from botocore.exceptions import ClientError
 from connectors.logger import logger, set_extra_logger
-
+from connectors.source import BaseDataSource
 
 SUPPORTED_CONTENT_TYPE = [
     "text/plain",
 ]
 SUPPORTED_FILETYPE = [".py", ".rst", ".rb", ".sh", ".md", ".txt"]
 ONE_MEGA = 1048576
+ENDPOINT_URL = os.getenv(
+    key="ENDPOINT_URL", default=None
+)  # For end to end tests set this environment variable with target host.
 
 
 class S3DataSource(BaseDataSource):
     """Amazon S3"""
 
     def __init__(self, connector):
+        """Setup connection to the Amazon S3.
+
+        Args:
+            connector (BYOConnector): Object of the BYOConnector class.
+        """
         super().__init__(connector)
         self.session = aioboto3.Session()
         self.loop = asyncio.get_event_loop()
         set_extra_logger(aws_logger, log_level=logging.INFO, prefix="S3")
         set_extra_logger("aioboto3.resources", log_level=logging.INFO, prefix="S3")
-
-    async def close(self):
-        await self.dl_client.close()
+        self.bucket_list = []
+        self.config = AioConfig(
+            read_timeout=self.configuration["read_timeout"],
+            connect_timeout=self.configuration["connect_timeout"],
+            retries={"max_attempts": self.configuration["max_attempts"]},
+        )
 
     async def ping(self):
-        return True
-
-    async def _get_content(self, key, timestamp):
-        # reuse the same for all files
-
-        logger.debug(f"Downloading {key}")
+        """Verify the connection with AWS"""
+        try:
+            async with self.session.client(
```

Comment:
you have a recurrent pattern in your code:

```
async with self.session.client(service_name="s3", config=self.config, endpoint_url=ENDPOINT_URL) as s3:
    # do something
```

let's create a `client` method in the class that is decorated with `contextlib.contextmanager` so your code can do:

```
async with self.client(**extra_options_if_needed) as s3:
    # do something
```